assets.go

```diff
@@ -1,354 +1,134 @@
-}
+// Formats manages resource handling of registered file formats.
+type Formats struct {
+	// formats maps from file extensions to resource loaders.
+	formats map[string]FileLoader
 
-func (l *Loader) Level(name string) *Level {
-	return l.levels[name]
+	// root is the directory which is prepended to every resource url internally
+	root string
 }
 
-func (l *Loader) Sound(name string) ReadSeekCloser {
-	f, err := os.Open(l.sounds[name])
-	if err != nil {
-		return nil
-	}
-	return f
+// SetRoot can be used to change the default directory from `assets` to whatever you want.
+//
+// Whenever `root` does not start with the directory `assets`, you will not be able to support mobile (Android/iOS)
+// since they require you to put all resources within the `assets` directory. You can, however, use subfolders within
+// the `assets` folder, and set those as `root`.
+func (formats *Formats) SetRoot(root string) {
+	formats.root = root
 }
 
-// ReadSeekCloser is an io.ReadSeeker and io.Closer.
-type ReadSeekCloser interface {
-	io.ReadSeeker
-	io.Closer
+// Register registers a resource loader for the given file format.
+func (formats *Formats) Register(ext string, loader FileLoader) {
+	formats.formats[ext] = loader
 }
 
-func (l *Loader) Load(onFinish func()) {
-	for _, r := range l.resources {
-		switch r.kind {
-		case "png":
-			if _, ok := l.images[r.name]; ok {
-				continue // with other resources
-			}
-
-			data, err := loadImage(r)
-			if err != nil {
-				log.Println("Error loading resource:", err)
-				continue // with other resources
-			}
-
-			l.images[r.name] = NewTexture(data)
-		case "jpg":
-			if _, ok := l.images[r.name]; ok {
-				continue // with other resources
-			}
-
-			data, err := loadImage(r)
-			if err != nil {
-				log.Println("Error loading resource:", err)
-				continue // with other resources
-			}
-
-			l.images[r.name] = NewTexture(data)
-		case "json":
-			if _, ok := l.jsons[r.name]; ok {
-				continue // with other resources
-			}
-
-			data, err := loadJSON(r)
-			if err != nil {
-				log.Println("Error loading resource:", err)
-				continue // with other resources
-			}
-
-			l.jsons[r.name] = data
-		case "tmx":
-			if _, ok := l.levels[r.name]; ok {
-				continue // with other resources
-			}
-
-			data, err := createLevelFromTmx(r)
-			if err != nil {
-				log.Println("Error loading resource:", err)
-				continue // with other resources
-			}
-
-			l.levels[r.name] = data
-		case "wav":
-			l.sounds[r.name] = r.url
-		case "ttf":
-			if _, ok := l.fonts[r.name]; ok {
-				continue // with other resources
-			}
-
-			f, err := loadFont(r)
-			if err != nil {
-				log.Println("Error loading resource:", err)
-				continue // with other resources
-			}
-
-			l.fonts[r.name] = f
+// Load loads the given resource into memory.
+func (formats *Formats) Load(url string) error {
+	ext := filepath.Ext(url)
+	if loader, ok := Files.formats[ext]; ok {
+		readCloser, err := openFile(filepath.Join(formats.root, url))
```

Comment:
From https://github.com/golang/go/wiki/CodeReviewComments#variable-names
> Variable names in Go should be short rather than long. This is especially true for local variables with limited scope. Prefer c to lineCount.

Suggestion: rename `readCloser` to `f`.